packages/material-ui-lab/src/TreeView/TreeView.js

```diff
@@ -98,31 +104,136 @@ const TreeView = React.forwardRef(function TreeView(props, ref) {
    */
 
   const getNextNode = (id) => {
-    const nodeIndex = visibleNodes.current.indexOf(id);
-    if (nodeIndex !== -1 && nodeIndex + 1 < visibleNodes.current.length) {
-      return visibleNodes.current[nodeIndex + 1];
+    // If expanded get first child
+    if (isExpanded(id) && getChildrenIds(id).length > 0) {
+      return getChildrenIds(id)[0];
+    }
+
+    // Try to get next sibling
+    const node = nodeMap.current[id];
+    const siblings = getChildrenIds(node.parentId);
+
+    const nextSibling = siblings[siblings.indexOf(id) + 1];
+
+    if (nextSibling) {
+      return nextSibling;
+    }
+
+    // try to get parent's next sibling
+    const parent = nodeMap.current[node.parentId];
+    if (parent) {
+      const parentSiblings = getChildrenIds(parent.parentId);
+      return parentSiblings[parentSiblings.indexOf(parent.id) + 1];
     }
     return null;
   };
 
   const getPreviousNode = (id) => {
-    const nodeIndex = visibleNodes.current.indexOf(id);
-    if (nodeIndex !== -1 && nodeIndex - 1 >= 0) {
-      return visibleNodes.current[nodeIndex - 1];
+    const node = nodeMap.current[id];
+    const siblings = getChildrenIds(node.parentId);
+    const nodeIndex = siblings.indexOf(id);
+
+    if (nodeIndex === 0) {
+      return node.parentId;
     }
-    return null;
+
+    let currentNode = siblings[nodeIndex - 1];
+    while (isExpanded(currentNode) && getChildrenIds(currentNode).length > 0) {
+      currentNode = getChildrenIds(currentNode).pop();
+    }
+
+    return currentNode;
   };
 
-  const getLastNode = () => visibleNodes.current[visibleNodes.current.length - 1];
-  const getFirstNode = () => visibleNodes.current[0];
-  const getParent = (id) => nodeMap.current[id].parent;
+  const getLastNode = () => {
+    let lastNode = getChildrenIds(null).pop();
+
+    while (isExpanded(lastNode)) {
+      lastNode = getChildrenIds(lastNode).pop();
+    }
+    return lastNode;
+  };
+  const getFirstNode = () => getChildrenIds(null)[0];
+  const getParent = (id) => nodeMap.current[id].parentId;
+
+  /**
+   * This is used to determine the start and end of a selection range so
+   * we can get the nodes between the two border nodes.
+   *
+   * It finds the nodes' common ancestor using
+   * a naive implementation of a lowest common ancestor algorithm
+   * (https://en.wikipedia.org/wiki/Lowest_common_ancestor) and
+   * then compares the ancestors below it that are also above the
+   * given nodes so we can compare their indexes to work out which
+   * node is higher in the tree view.
+   */
+  const getOrderOfNodesInTree = (nodeAId, nodeBId) => {
```

Comment:
Other attempt? This returns the nodes ordered by their depth in the tree?

Reply:
@joshwooding Was my explanation correct? Seems like this is method is overly complicated. Wouldn't it make more sense to move a pointer from both nodes to its parent and the first pointer that reaches the root (or once a pointer is moved to a node that was already visited) determines in which order the nodes are returned?

Reply:
Depth could be mistaken for actual depth without taking into consideration the check looks for which node is the most right/left

Reply:
> @joshwooding Was my explanation correct? Seems like this is method is overly complicated. Wouldn't it make more sense to move a pointer from both nodes to its parent and the first pointer that reaches the root (or once a pointer is moved to a node that was already visited) determines in which order the nodes are returned?

What happens when they both have the same parent?

Reply:
> This method also tries to avoid going all the way up the tree but it essentially does what you've said and checks the indices.

Still O(tree-depth) :P

You should explain more how you define "order" here. Order is a generic term or even worse: Order is something you need to defined for it to make sense.

Maybe this is the order in which the nodes are encountered when walking the tree depth-first?

Reply:
Exactly that my new go reads like this:

  /**
   * This is used to determine the start and end of a selection range so
   * we can get the nodes between the two border nodes.
   *
   * It finds the nodes' common ancestor using
   * a naive implementation of a lowest common ancestor algorithm
   * (https://en.wikipedia.org/wiki/Lowest_common_ancestor).
   * Then compares the ancestor's 2 children that are ancestors of nodeA and NodeB
   * so we can compare their indexes to work out which node comes first in a depth first search.
   * (https://en.wikipedia.org/wiki/Depth-first_search)
   *
   * Another way to put it is which node is deeper in a trémaux tree
   * https://en.wikipedia.org/wiki/Tr%C3%A9maux_tree
   */

Reply:
Yep makes sense. We'll see when we touch the code the next time if it still holds :smile: